src/main/scala/is/hail/io/bgen/BgenRDDPartitions.scala

```diff
@@ -60,53 +54,48 @@ object BgenRDDPartitions extends Logging {
       Array.empty
     } else {
       val partitions = new ArrayBuilder[Partition]()
-      var partitionIndex = 0
       var fileIndex = 0
       while (fileIndex < nonEmptyFilesAfterFilter.length) {
         val file = nonEmptyFilesAfterFilter(fileIndex)
         val nPartitions = fileNPartitions(fileIndex)
-        val maxRecordsPerPartition = (file.nVariants + nPartitions - 1) / nPartitions
         using(new OnDiskBTreeIndexToValue(file.path + ".idx", hConf)) { index =>
           includedVariantsPerFile.get(file.path) match {
             case None =>
-              val startOffsets =
-                index.positionOfVariants(
-                  Array.tabulate(nPartitions)(i => i * maxRecordsPerPartition))
+              val partNVariants = partition(file.nVariants, nPartitions)
+              val partFirstVariantIndex = partNVariants.scan(0)(_ + _).init
+              val partFirstByteOffset = index.positionOfVariants(partFirstVariantIndex)
               var i = 0
               while (i < nPartitions) {
                 partitions += BgenPartitionWithoutFilter(
                   file.path,
                   file.compressed,
-                  i * maxRecordsPerPartition,
-                  startOffsets(i),
-                  if (i < nPartitions - 1) startOffsets(i + 1) else file.fileByteSize,
-                  partitionIndex,
```

Comment:
`i` is the index of the partition in this file rather than a global partition index over all the files that might be being loaded. Is that OK?